nilmtk/api.py

```diff
@@ -141,14 +141,15 @@ def train_chunk_wise(self, clf, d, current_epoch):
                         train_df = pd.DataFrame(np.zeros(appliance_readings[0].shape),index = appliance_readings[0].index,columns=appliance_readings[0].columns)
                         for app_reading in appliance_readings:
                             train_df+=app_reading
-                    train_appliances = []
 
-                    for cnt,i in enumerate(appliance_readings):
-                        train_appliances.append((self.appliances[cnt],[i]))
+                    for i,appliance_name in enumerate(self.appliances):
+                        if appliance_name in self.train_data:
+                            self.train_data[appliance_name]["mains"].append(train_df)
+                            self.train_data[appliance_name]["appliance"].append(appliance_readings[i])
+                        else:
+                            self.train_data[appliance_name] = { "mains" : [train_df], "appliance" : [appliance_readings[i]]}
 
-                    self.train_mains = [train_df]
-                    self.train_submeters = train_appliances
-                    clf.partial_fit(self.train_mains, self.train_submeters, current_epoch)
+                    clf.partial_fit(self.train_data, current_epoch)
```

Comment:
This change breaks the compatibility with NILMTK's `Disaggregator` class and with all implementations of `nilmtk_contrib`.
What is your rationale for such a big breaking change?

Reply:
The reason behind such change is the fact that when using appliance activations, the periods of readings for one appliance are different from the other appliances.
This requires identifying which aggregated readings are relevant for the given appliance as we are not using the entire period and instead only some subsets.

I knew that the changes would break the Disaggregator class, but I decided it was better to make a pull request and debate the best way to add such changes to the API and Disaggregator class.

Reply:
Happy to discuss a technical solution with you! My goal here is to allow `nilmtk` to evolve and fit the needs without breaking for everyone else.

I am still not sure to understand everything. Could you please update your PR description with a picture showing how you change segmentation from the current sliding window?

Reply:
So the need to use a dictionary is directed by the fact that the different appliances will have different activations.
With different activations, the relevant readings of the mains data will also be different ,
The figure below might help explain the proposal and why we need to change methods' signatures.

![temp](https://user-images.githubusercontent.com/43450638/143453026-a8c40c91-2478-466a-a468-f3f4c57a065c.png)

--------------------------------
In my implementation, I also divided the training data into train and cross-validation data,
allowing the user to specify, for example, different houses for training/cross-validating the model.
But I was only going to propose it later.

Reply:
Thank you for the explanation. I understand you want a different resolution for each target appliance.
A few questions:
1. Why not just using a different `sequence_length` for each disaggregator?
2. Do you want to omit windows with no activation of the target appliance while training?
3. For segmentation (event detection), do you use the input (aggregated) signal or the output (disaggregated) signal?
4. How would your method be applied to an unseen home, without knowing the ground truth?

Reply:
1. By using a different sequence_length, we are not solving the problem where the microwave, for example, is off 98% of the day. So, our models will be biased towards the off category instead of correctly learning to detect.
2. Yes, exactly. The idea here is to follow what was proposed in NeuralNILM and some other papers.
3/4. This new method aims to obtain a balanced training/testing set to evaluate the models so we can use the ground truth of the appliances. In a real-life scenario, we would disaggregate every timestep in the time series